src/kOS.Safe/Utilities/KOSNomenclature.cs

```diff
@@ -0,0 +1,184 @@
+﻿using System;
+using System.Text;
+using System.Reflection;
+using System.Collections.Generic;
+using System.Linq;
+using kOS.Safe.Encapsulation;
+using kOS.Safe.Exceptions;
+using kOS.Safe.Serialization;
+
+namespace kOS.Safe.Utilities
+{
+    /// <summary>
+    /// A class for the purpose of holding the mapping information that goes
+    /// back and forth between C# type names and thier KOS type name equivalents.
+    /// For the sake of this we are using unqualified C# names (without the namespace
+    /// prefixes).
+    /// 
+    /// Any time you add a new class derived from kOS.Safe.Encapsulated.Structure, you
+    /// should check to see if you should add a mapping to this class to go with it.
+    /// 
+    /// This class is intended to be used "statically" without an actual instance of it
+    /// being necessary.  There should be only one copy of its data, globally, across
+    /// the entire process.
+    /// </summary>
+    public class KOSNomenclature
+    {
+        private static HashSet<Assembly> seenAssemblies = new HashSet<Assembly>();
+        private static Dictionary<string,Type> kosToCSharpMap = new Dictionary<string, Type>(StringComparer.OrdinalIgnoreCase);
+        private static Dictionary<Type,string> cSharpToKosMap = new Dictionary<Type, string>();
+
+        private static string NagHeader =
+            "***********************************************\n" +
+            "* The following C# types need a KOS name      *\n" +
+            "* mapping defined in the C# code by use of a  *\n" +
+            "* KOSNomenclatureAttribute.  If you see this  *\n" +
+            "* message it means a KOS developer added a    *\n" +
+            "* new type without finishing the job.         *\n" +
+            "* If you _ARE_ a kOS developer and see this   *\n" +
+            "* message after you've just added a new class *\n" +
+            "* then you need to fix this.                  *\n" +
+            "***********************************************\n";
+
+        // Neither public nor private because a static constructore can't be called explicitly anyway.
+        static KOSNomenclature()
+        {
+        }
+        
+        public static void PopulateMapping(params Assembly[] assemblies)
+        {
+            foreach (Assembly a in assemblies)
+            {
+                // Ensure an Assembly is never run through this twice in the life of the program:
+                if (seenAssemblies.Contains(a))
+                    return;
```

Comment:
Good point.  Hopefully I'll remember to go back in and fix that later when I get back to this in several hours.